server/rest/src/main/java/org/infinispan/rest/resources/ProtobufResource.java

```diff
@@ -0,0 +1,233 @@
+import org.infinispan.rest.InvocationHelper;
+import org.infinispan.rest.NettyRestResponse;
+import org.infinispan.rest.cachemanager.RestCacheManager;
+import org.infinispan.rest.framework.ContentSource;
+import org.infinispan.rest.framework.ResourceHandler;
+import org.infinispan.rest.framework.RestRequest;
+import org.infinispan.rest.framework.RestResponse;
+import org.infinispan.rest.framework.impl.Invocations;
+import org.infinispan.rest.operations.exceptions.NoDataFoundException;
+import org.infinispan.rest.operations.exceptions.NoKeyException;
+
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.infinispan.commons.dataconversion.MediaType.APPLICATION_JSON;
+import static org.infinispan.rest.framework.Method.DELETE;
+import static org.infinispan.rest.framework.Method.GET;
+import static org.infinispan.rest.framework.Method.POST;
+import static org.infinispan.rest.framework.Method.PUT;
+
+/**
+ * Protobuf schema manipulation Resource
+ *
+ * @author Katia Aresti
+ * @since 11
+ */
+public class ProtobufResource extends BaseCacheResource implements ResourceHandler {
+
+   private final ObjectMapper objectMapper;
+
+   public ProtobufResource(InvocationHelper invocationHelper) {
+      super(invocationHelper);
+      objectMapper = invocationHelper.getMapper();
+   }
+
+   @Override
+   public Invocations getInvocations() {
+      return new Invocations.Builder()
+            // Key related operations
+            .invocation().methods(GET).path("/v2/schema").handleWith(this::getSchemasNames)
+            .invocation().methods(POST).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,true))
+            .invocation().methods(PUT).path("/v2/schema/{schemaName}").handleWith(r -> addOrEditSchema(r,false))
+            .invocation().methods(GET).path("/v2/schema/{schemaName}").handleWith(this::getSchema)
+            .invocation().method(DELETE).path("/v2/schema/{schemaName}").handleWith(this::deleteSchema)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> getSchemasNames(RestRequest request) {
+      NettyRestResponse.Builder responseBuilder = new NettyRestResponse.Builder();
+
+      AdvancedCache<Object, Object> cache = invocationHelper.getRestCacheManager()
+            .getCache(ProtobufMetadataManager.PROTOBUF_METADATA_CACHE_NAME, request);
+
+
+      return CompletableFuture.supplyAsync(() ->
+          Flowable.fromIterable(cache.entrySet())
+               .filter(cacheEntry -> !((String)cacheEntry.getKey()).endsWith(ProtobufMetadataManager.ERRORS_KEY_SUFFIX))
+               .map(cacheEntry -> {
+                  String error = (String) cache
+                        .get(cacheEntry.getKey() + ProtobufMetadataManagerConstants.ERRORS_KEY_SUFFIX);
```

Comment:
This is another blocking call here. Unfortunately, the more I look at this it should really be done using BlockingManager#blockingPublisher on the `Flowable.fromIterable(cache.entrySet())` and then doing non blocking calls aftwerds.

Reply:
But I guess we can leave this as is for now. This whole chunk will need to be redone to use a ClusterPublisher at some point anyways.